specs changes for azure databox ga

sojain opened the pull request:
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

Currently raised against preview, once signed off will move it to stable

soumya-jain19 committed:
specs changes for azure databox ga

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-ruby

AutorestCI commented:
# Automation for azure-sdk-for-java

AutorestCI commented:
# Automation for azure-sdk-for-node

AutorestCI commented:
# Automation for azure-sdk-for-go

azuresdkci commented:
Can one of the admins verify this patch?

anuchandy commented on specification/databox/resource-manager/Microsoft.DataBox/preview/2018-01-01/databox.json:
```diff
@@ -1932,113 +2063,133 @@
+            "name": "SkuDisabledReason",
+            "modelAsString": false
+          }
+        },
```
You can always provide description of enum values as below:

```json
        "disabledReason": {
          "type": "string",
          "description": "Reason why the Sku is disabled.",
          "enum": [
            "None",
            "Country",
            "Region",
            "Feature",
            "OfferType",
            "Bitrate"
          ],
          "x-ms-enum": {
            "name": "SkuDisabledReason",
            "values": [
              {
                "value": "None",
                "description": "disabled due to unknown reason"
              },
              {
                "value": "Country",
                "description": "SKU is not available in the country."
              },
              {
                "value": "Feature",
                "description": "Feature is not enabled for the SKU."
              },
              ...
            ],
            "modelAsExtensible": false
          },
        },
```

sojain replied:
nice to learn comment taken

anuchandy reviewed (commented):
in general modelAsString:false has extensibility constrain, i.e. if service introduce a new value for the enum in the same api-version then SDKs can break. If we are not anticipating any new values in the current api-version then we are good here otherwise keep `modelAsString:false`. (for the discriminator specific enums you can keep `modelAsString:false`).

anuchandy commented:
@ravbhatnagar please take a look, this is addition to preview feature. Also need your input on linked access refer [this](
https://github.com/Azure/azure-rest-api-specs/pull/3885#discussion_r217202024).

anuchandy commented:
@sojain between you will need to update your readme file to have tags, like below:


# DataBox

> see https://aka.ms/autorest

This is the AutoRest configuration file for DataBox.



---
## Getting Started
To build the SDK for DataBox, simply [Install AutoRest](https://aka.ms/autorest/install) and in this folder, run:

> `autorest`

To see additional help and options, run:

> `autorest --help`
---

## Configuration

### Basic Information
These are the global settings for the DataBox API.

``` yaml
openapi-type: arm
tag: package-2018-01
```

### Tag: package-2018-01

These settings apply only when `--tag=package-2018-01` is specified on the command line.

``` yaml $(tag) == 'package-2018-01'
input-file:
- Microsoft.DataBox\preview\2018-01-01\databox.json
```

## Suppression

``` yaml
directive:
  - suppress:
    - R2016 #to suppress (PatchBodyParametersSchema/R2016/RPCViolation)
    - R2062 #to suppress (XmsResourceInPutResponse/R2062/RPCViolation)
```

---
# Code Generation


## Swagger to SDK

This section describes what SDK should be generated by the automatic system.
This is not used by Autorest itself.

``` yaml $(swagger-to-sdk)
swagger-to-sdk:
  - repo: azure-sdk-for-go
  - repo: azure-sdk-for-node
```


## C#

These settings apply only when `--csharp` is specified on the command line.
Please also specify `--csharp-sdks-folder=<path to "SDKs" directory of your azure-sdk-for-net clone>`.

``` yaml $(csharp)
csharp:
  azure-arm: true
  license-header: MICROSOFT_MIT_NO_VERSION
  namespace: Microsoft.Azure.Management.DataBox
  output-folder: $(csharp-sdks-folder)/DataBox/Management.DataBox/Generated
  clear-output-folder: true
```


## Go

These settings apply only when `--go` is specified on the command line.

``` yaml $(go)
go:
  license-header: MICROSOFT_APACHE_NO_VERSION
  namespace: databox
  clear-output-folder: true
```

### Go multi-api

``` yaml $(go) && $(multiapi)
batch:
  - tag: package-2018-01
```

### Tag: package-2018-01 and go

Please also specify `--go-sdk-folder=<path to the root directory of your azure-sdk-for-go clone>`.

``` yaml $(go)
output-folder: $(go-sdk-folder)/services/preview/databox/mgmt/2018-01-01/databox
```


## Java

These settings apply only when `--java` is specified on the command line.
Please also specify `--azure-libraries-for-java-folder=<path to the root directory of your azure-libraries-for-java clone>`.

``` yaml $(java)
java:
  azure-arm: true
  fluent: true
  namespace: com.microsoft.azure.management.databox
  license-header: MICROSOFT_MIT_NO_CODEGEN
  payload-flattening-threshold: 1
  output-folder: $(azure-libraries-for-java-folder)/databox
```

### Java multi-api

``` yaml $(java) && $(multiapi)
batch:
  - tag: package-2018-01
```

### Tag: package-2018-05 and java

These settings apply only when `--tag=package-2018-01 --java` is specified on the command line.
Please also specify `--azure-libraries-for-java-folder=<path to the root directory of your azure-sdk-for-java clone>`.

``` yaml $(tag)=='package-2018-01' && $(java) && $(multiapi)
java:
  namespace: com.microsoft.azure.management.databox.v2018_01_01
  output-folder: $(azure-libraries-for-java-folder)/databox/resource-manager/v2018_01_01
regenerate-manager: true
generate-interface: true
```

sojain commented on specification/databox/resource-manager/Microsoft.DataBox/preview/2018-01-01/databox.json:
```diff
@@ -1145,22 +1002,295 @@
+        "accountId"
+      ],
+      "type": "object",
+      "properties": {
```
We have enabled linked access to storage service, and in most happy paths we only take accountId from customer thats why it is the only required field. Customers came up ith requirements that they want to copy data to National clouds, though we dont have services in that region yet and linked access from public cloud will not work on national cloud. Hence it became important to give customers the option to pass these parameters , can you suggest ways i can make it more explicit in comments

sojain replied:
Post more discussion with leads we have removed these changes, we will give a private sdk for handful of national cloud customers

sojain commented on specification/databox/resource-manager/Microsoft.DataBox/preview/2018-01-01/databox.json:
```diff
@@ -1765,26 +1816,108 @@
+        "status": {
+          "description": "Status of the service.",
+          "type": "boolean"
+        }
```
These properties are not used will remove the same

sojain commented on specification/databox/resource-manager/Microsoft.DataBox/preview/2018-01-01/databox.json:
```diff
@@ -1765,26 +1816,108 @@
+          "description": "Version of the service.",
+          "type": "string"
+        }
+      }
```
These properties are not used will remove the same

sojain commented on specification/databox/resource-manager/Microsoft.DataBox/preview/2018-01-01/databox.json:
```diff
@@ -802,6 +670,20 @@
+          "description": "Mac Address",
+          "type": "string"
+        }
+      }
```
sure, will do

KrisBash commented on specification/databox/resource-manager/Microsoft.DataBox/preview/2018-01-01/databox.json:
```diff
@@ -536,103 +555,19 @@
-        }
-      }
-    },
-    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataBox/jobs/{jobName}/copyLogsUri": {
```
What is the migration path for deprecation of this API? It appears to be in active current use.

sojain replied:
Till now are only active user is our own portal and we have stopped using these apis, our production systems have already stopped supporting these and see no usage/failure because of its remove. Hence as we go GA on 24th september we dont want the apis to be available for our GA sdk

sojain replied:
Here is the plan
1) For Apis GetCopyLogsUri, DownloadShippingLabelUri, ReportIssue have zero users and the api has been disabled and non exisitent in our service for more than a month, so we are removing it from sdk also
2) ListSecrets is continued to be supported from service but is no longer support in the new generated client dll, anyone calling the api will be returned a proper response

Just FYI we have only testing (internal) sdk users, post we go GA we will have more sdk users

KrisBash replied:
@sojain There are no external users of this API? Why not take this to a new API version? These are significant changes in an API version. Are you planning to move to a stable API for your GA?

sojain replied:
@KrisBash , yes this pr is to show the diffs, as soon as I have a sign off ill update this spec to stable and only post its completion we will generate a new dll for customer

KrisBash replied:
@sojain Thanks

KrisBash reviewed (changes requested):
Even though this is a preview version, these changes need to correspond to a new API version to avoid breaking customers. Some of these deleted POST apis are actively being used, and a deprecation plan is advised.

soumya-jain19 committed:
code review comments
1) Removing unnecessary fields
2) Correctly populating modelasstring as modelasextensible
3) adding descriptions for enum

sojain commented on specification/databox/resource-manager/Microsoft.DataBox/preview/2018-01-01/databox.json:
```diff
@@ -855,64 +737,22 @@
     },
```
taken appropriate fixes

soumya-jain19 committed:
fixes to readme.md and readme.nodejs.md

sojain commented:
@anuchandy , i have taken most of the comments and also the readme update, however post that build is not working, can you help me in finding the root cause

soumya-jain19 committed:
1) Fixing readme.md for build
2) Moving to stable folder

anuchandy commented:
@sojain I've created readme file for databox here https://raw.githubusercontent.com/anuchandy/sample-readme/master/readme.md which includes settings for all languages. Can you try this?

anuchandy commented:
@sojain you also need sdks for all languages right? the readme contains entries for C# and Go only. The [link](
https://raw.githubusercontent.com/anuchandy/sample-readme/master/readme.md) i shared contains entries for all languages.

soumya-jain19 committed:
updating readme to include sdk for all languages

sojain commented:
@anuchandy , thanks for sharing this.

This failed with 
_Command: bundle install && rake arm:regen_all_profiles['azure_mgmt_databox']
Finished with return code 10
and output:
Don't run Bundler as root. Bundler can ask for sudo if it is needed, and
installing your bundle as root will break this application for all non-root
users on this machine.
Could not locate Gemfile_

It works if a remove bundle command, any suggestions. Also just fo rmy knowledge did you some tool to generate the readme.md

anuchandy commented:
@lmazuel the readme's swagger-to-sdk section looks like below:

```json
swagger-to-sdk:
  - repo: azure-sdk-for-python
  - repo: azure-sdk-for-node
  - repo: azure-sdk-for-go
  - repo: azure-sdk-for-ruby
  - repo: azure-sdk-for-java
    after_scripts:
      - bundle install && rake arm:regen_all_profiles['azure_mgmt_databox']
```

The command `bundle install && rake arm:regen_all_profiles['azure_mgmt_databox']` to generate the ruby sdk is failing with error 

```
Command: bundle install && rake arm:regen_all_profiles['azure_mgmt_databox']
Finished with return code 10
and output:
Don't run Bundler as root. Bundler can ask for sudo if it is needed, and
installing your bundle as root will break this application for all non-root
users on this machine.
Could not locate Gemfile
```

Talked to @sarangan12 about this, he mentioned that this command shouldn't be run as a root user. Can you take a look and help here?

anuchandy commented:
@sojain i manually author the readme, it's not generated.

sarangan12 commented:
@anuchandy The error is expected. It is not because of the root user. The azure_mgmt_databox is a new gem which does not have a skeleton structure. So, it is failing. You can ignore this error

sojain commented:
thanks @anuchandy for all the help, who can help me in closing this pull request given I have arm signoff

soumya-jain19 committed:
Addition of optional list of sku names for filter while getting skuinfo

anuchandy commented:
@sojain - is this api-version deployed to at least one production region? if yes, we're good to merge. let me know.

sojain commented:
@anuchandy yes api version is deployed to all regions

sojain commented:
@anuchandy, sorry to bother, but can you please help in completing this pull request, we have GA at ignite on 24th and hence the urgency

AutorestCI commented:
# Automation for azure-sdk-for-js
